src/Symfony/Component/DependencyInjection/Compiler/PriorityTaggedServiceTrait.php

```diff
@@ -50,82 +52,68 @@ private function findAndSortTaggedServices($tagName, ContainerBuilder $container
+                    if (!($rm = $r->getMethod($defaultIndexMethod))->isStatic()) {
+                        throw new InvalidArgumentException(sprintf('Method "%s::%s()" should be static: tag "%s" on service "%s" is missing "%s" attribute.', $class, $defaultIndexMethod, $tagName, $serviceId, $indexAttribute));
+                    }
+                    if (!$rm->isPublic()) {
+                        throw new InvalidArgumentException(sprintf('Method "%s::%s()" should be public: tag "%s" on service "%s" is missing "%s" attribute.', $class, $defaultIndexMethod, $tagName, $serviceId, $indexAttribute));
+                    }
 
-                continue;
-            }
+                    $defaultIndex = $rm->invoke(null);
 
-            if (!$r && !$r = $container->getReflectionClass($class)) {
-                throw new InvalidArgumentException(sprintf('Class "%s" used for service "%s" cannot be found.', $class, $serviceId));
+                    if (!\is_string($defaultIndex)) {
+                        throw new InvalidArgumentException(sprintf('Method "%s::%s()" should return a string, got %s: tag "%s" on service "%s" is missing "%s" attribute.', $class, $defaultIndexMethod, \gettype($defaultIndex), $tagName, $serviceId, $indexAttribute));
+                    }
+                }
             }
 
-            $class = $r->name;
-
-            if (!$r->hasMethod($defaultIndexMethod)) {
-                if ($needsIndexes) {
-                    $services[$priority][$serviceId] = new TypedReference($serviceId, $class);
+            foreach ($attributes as $attribute) {
+                $priority = isset($attribute['priority']) ? $attribute['priority'] : $defaultPriority;
+                $name = $index = $indexAttribute && isset($attribute[$indexAttribute]) ? $attribute[$indexAttribute] : $defaultIndex;
 
-                    continue;
+                if (null === $index && $needsIndexes) {
+                    $index = $serviceId;
                 }
 
-                throw new InvalidArgumentException(sprintf('Method "%s::%s()" not found: tag "%s" on service "%s" is missing "%s" attribute.', $class, $defaultIndexMethod, $tagName, $serviceId, $indexAttribute));
-            }
-
-            if (!($rm = $r->getMethod($defaultIndexMethod))->isStatic()) {
-                throw new InvalidArgumentException(sprintf('Method "%s::%s()" should be static: tag "%s" on service "%s" is missing "%s" attribute.', $class, $defaultIndexMethod, $tagName, $serviceId, $indexAttribute));
-            }
-
-            if (!$rm->isPublic()) {
-                throw new InvalidArgumentException(sprintf('Method "%s::%s()" should be public: tag "%s" on service "%s" is missing "%s" attribute.', $class, $defaultIndexMethod, $tagName, $serviceId, $indexAttribute));
-            }
+                $reference = $r && 'stdClass' !== $class ? new TypedReference($serviceId, $class, $behavior, $name) : new Reference($serviceId, $behavior);
```

Comment:
why is stdClass a special case? what's the relation with the fix?

Reply:
stdClass should not be TypedReference as they are not actual defined types.